RFC-0011-coordinator-throttling.md

```diff
@@ -0,0 +1,117 @@
+- Also decrease overall variance of execution time of queries (queuing is fine)
+
+## Proposed Implementation
+### Worker side changes
+- Added endpoint to gather worker load (`/v1/info/nodestats`)
+  - Include NodeState + worker load. Idea is to just call `/v1/info/nodestats` rather than `/v1/info/state` (eventually deprecate it). Worker load is already getting populated in [cpp workers](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/PrestoServer.cpp#L1537).  
+- PR: https://github.com/prestodb/presto/pull/25686
+
+### Coordinator side changes
+#### Collecting worker load data
+  - Update DiscoveryManager to invoke this end point rather than the `/v1/info/state`
+  - Expose `getNodeLoadMetrics` on InternalNodeManager
+  - PR: https://github.com/prestodb/presto/pull/25688 
+#### Scheduling policies
+  - Implement node overload policies that can govern the policies to determine cluster overload
+  - Based on these policies and load collected from worker, improve the admission control logic globally to not queue the query (irrespective of Resource Group)
+  - PR: https://github.com/prestodb/presto/pull/25689 
+
+
+## Metrics
+
+- Cluster overload
+- Variance in runtime of queries
+
+
+## Other Approaches Considered
+
+### Approach 1
+#### Improve Granular task scheduling
+Not fully baked, but at high-level based on load, the coordinator can assign scores to the workers and selects the top N highest-scoring workers, then randomly chooses from this subset to schedule tasks or splits.
+
+**Few Examples**:
+- **Intermediate stage**: For 333 hash partitions, randomly select 333 workers from the top 500 scoring workers. (We run with 300 node clusters so this property needs to be tuned as well if we go with this approach.)
+- **Leaf stage**: For leaf task, the logic in `chooseLeastBusyNode` could be augmented to pick workers with best score and out of these assign the splits.
+
+
+- **Pros**
+This could provide load balancing to begin with and prevent some workers getting hot
+
+- **Cons**
+In our cluster most of the workers become overloaded in a fraction of minutes due to a lot of queries getting admitted. 
+
+Load balancing may not be immediately helpful in such cases
+This could be a good direction for the future if we see only a select few workers are causing issues.
+
+### Approach 2
+Write a separate service which can consume worker load and update global RG (max queries that can run based on the load / policies)
+
+- **Pros**
+No coordinator changes
+
+- **Cons**
+Extra service to maintain
+Two brains
+
+
+## Adoption Plan
+If this feature is enabled, this can cause queuing when cluster is loaded (which would be separate from queuing due to Resource Groups)
+. For now end users can correlate queuing via looking at queuing + worker overload metrics together. If required, we can also expose separate queuing metrics due to cluster overload.
+## Test Plan
+Added the test cases
```

Comment:
Spell this out more about what kinds of tests you are adding.

Reply:
Added the details on testing.